src/eth_provider/provider.rs

```diff
@@ -663,6 +659,23 @@ where
         Ok(Some(block.into()))
     }
 
+    /// Convert the given block id into a Starknet block id
+    async fn to_starknet_block_id(
+        &self,
+        block_id: impl Into<Option<BlockId>>,
+    ) -> EthProviderResult<starknet::core::types::BlockId> {
+        match block_id.into() {
+            Some(BlockId::Hash(hash)) => {
+                Ok(starknet::core::types::BlockId::Number(self.hash_into_block_number(hash).await?.to::<u64>()))
```

Comment:
why do we convert from hash to number?

Reply:
We can't query a starknet block by an evm hash, since a U256 doesn't fit in a Felt. The idea here is to convert the requested hash to a number, which can then be used for example to get the code of a contract for a particular block number (even though it was requested by hash).

Reply:
ok ok but two options:
- if the hash exists in kakarot, then hash should be expressed as u256 but is u251 -> we can check leading zeroes and then convert types?
- if the hash does not exist (it is a "true" evm block hash -> it is non existent in our db) -> we just return an error

Reply:
I think this is more work, the current flow is:
- search the db for the block_number for this hash. Make the rest of the query.
- if the hash is a true evm hash, it won't be in the db and return a `not found` error.
I could try to convert to U251, but what happens if it is an EVM hash, just fits in a U251? Then you lose time on the query, instead of just returning fast.

Reply:
alright